waitUntilReady now first waits until resource exists

Ladicek opened the pull request:
I'm submitting this only to discuss the approach I've taken. The code contains a couple of TODOs that need to be resolved first.

Also the PR doesn't contain any test; could you please advise how to approach testing this change? I'm not at all familiar with tests in this codebase.

Fixes #921.

Ladicek commented:
Aha, there's a test failure that also points me to a place where things like this are tested. Nice :-)

Ladicek commented:
Fixed failing tests and added a new test.

Ladicek commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/PodOperationsImpl.java:
```diff
@@ -335,22 +335,5 @@ public ExecWatch exec(String... command) {
-    try (Watch watch = watch(watcher)) {
-      return watcher.await(amount, timeUnit);
-    }
-  }
```
Just thinking -- not sure why `Pod.waitUntilReady` was different than `<everything else>.waitUntilReady`, but the common implementation seems to work just fine.

iocanel replied:
see my comment above.

Ladicek commented:
Ping @iocanel, did you maybe have time to take a look at this?

iocanel commented:
@Ladicek: no! I'll try to have a look at it during the weekend.

iocanel commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:
```diff
@@ -954,8 +954,34 @@ public Boolean isReady() {
+        return item;
+      }
+
+      // TODO is there a better way?
```
While implementing the `waitUntilReady` I found that the optimal approach is to go hybrid, by combining a while/wait loop and watchers. This limits the number of request send to the server, but also addresses potential races that can affect a purely watcher approach.

For me going with a pure while/wait just for the exists check is good enough at least for now.

iocanel commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/HasMetadataOperation.java:
```diff
@@ -211,4 +211,20 @@ protected T periodicWatchUntilReady(int i, long started, long interval, long amo
+      long alreadySpent = System.currentTimeMillis() - started;
+
+      // if awaiting existence took very long, let's give at least 10 seconds to awaiting readiness
+      long remaining = Math.max(10_000, timeUnit.toMillis(amount) - alreadySpent);
```
I am not sure if I fancy giving it extra time. Some possible alterations:

- subtract that 10 seconds from the waitUntilExists time.
- check if remaining is less or equal to zero and then just check if resource is ready.

iocanel commented on kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/PodOperationsImpl.java:
```diff
@@ -335,22 +335,5 @@ public ExecWatch exec(String... command) {
         return new PodOperationsImpl(client, getConfig(), apiVersion, namespace, name, isCascading(), getItem(), getResourceVersion(), isReloadingFromServer(), getGracePeriodSeconds(), getLabels(), getLabelsNot(), getLabelsIn(), getLabelsNotIn(), getFields(), containerId, in, inPipe, out, outPipe, err, errPipe, withTTY, withTerminatedStatus, withTimestamps, sinceTimestamp, sinceSeconds, withTaili ...
     }
-
-  @Override
```
This must have slipped. But its now removed so all fine.

Ladicek commented:
I've rebased on top of current `master` and added one commit to address the remaining items (removed the TODOs, better waiting for readiness when existence await took too long).

iocanel commented:
[merge]
